x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/MachineLearningPackageLoader.java

```diff
@@ -52,4 +55,52 @@ public List<Setting<?>> getSettings() {
             new ActionHandler<>(LoadTrainedModelPackageAction.INSTANCE, TransportLoadTrainedModelPackage.class)
         );
     }
+
+    @Override
+    public List<BootstrapCheck> getBootstrapChecks() {
+        return List.of(new BootstrapCheck() {
+            @Override
+            public BootstrapCheckResult check(BootstrapContext context) {
+                try {
+                    validateModelRepository(MODEL_REPOSITORY.get(context.settings()), context.environment().configFile());
+                } catch (Exception e) {
+                    return BootstrapCheckResult.failure(
+                        "Found an invalid configuration for xpack.ml.model_repository. "
+                            + e.getMessage()
+                            + ". See <TODO: URL> for more information."
+                    );
+                }
+                return BootstrapCheckResult.success();
+            }
+
+            @Override
+            public boolean alwaysEnforce() {
+                return true;
+            }
+        });
+    }
+
+    static void validateModelRepository(String repository, Path configPath) throws URISyntaxException {
+        URI baseUri = new URI(repository.endsWith("/") ? repository : repository + "/").normalize();
+        String normalizedConfigPath = configPath.normalize().toAbsolutePath().toString();
+
+        if (Strings.isNullOrEmpty(baseUri.getScheme())) {
+            throw new IllegalArgumentException(
+                "xpack.ml.model_repository must contain a scheme, possible schemes are \"http\", \"https\", \"file\""
+            );
+        }
+
+        final String scheme = baseUri.getScheme().toLowerCase(Locale.ROOT);
+        if (Set.of("http", "https", "file").contains(scheme) == false) {
+            throw new IllegalArgumentException(
+                "xpack.ml.model_repository must be configured with one of the following schemes: \"http\", \"https\", \"file\""
+            );
+        }
+
+        if (scheme.equals("file") && (baseUri.getPath().startsWith(normalizedConfigPath) == false)) {
+            throw new IllegalArgumentException(
+                "If xpack.ml.model_repository is a file location, it must be placed below the configuration path: " + configPath
```

Comment:
```suggestion
                "If xpack.ml.model_repository is a file location, it must be placed below the configuration path: " + normalizedConfigPath
```

I wonder if checking against the normalized path will be a can of worms on Windows - I am not sure what Java considers normalized on Windows. Possibly we'll need to replace backslashes with forward slashes before validating. However, since we're short of time I'm happy to leave this for 8.9.

But at least the error should be clear about exactly what needs to go in the URL to get the check to pass.

Reply:
I took the unnormalized path on purpose, because that's what the user knows about. The normalized form might be confusing.

Fully agree to the windows concerns, surprisingly java has not much to offer w.r.t. platform agnostic path handling (compared to the rich `os.path` APIs in python for example).

Anyway, I don't have a windows box, so it seems we have to jump into cold water and try it. When I merge I will watch for CI errors and mute or fix the tests.

Reply:
> I took the unnormalized path on purpose, because that's what the user knows about.

In that case that’s what we should compare to the URL, because then a user who sees that error knows exactly what to put in the URL to fix the error.

Reply:
hmm, actually `file://` is always absolute. `toAbsolutePath` is probably superflous

So here are examples from `elasticsearch.yml`:

```
xpack.ml.model_repository: file://${path.home}/config/models/
#xpack.ml.model_repository: file:///home.elk/elasticsearch/config/models/
#xpack.ml.model_repository: file://${path.home}/con/models/
#xpack.ml.model_repository: https://my-custom-backend
```

Windows should look like this?
```
#xpack.ml.model_repository: file:///c:/elk/config
```

so I think we are good regarding backslashes.

`normalize` just resolves dots. 

Anyway, I think we overcomplicate it, I will change it as you suggest. Lets see how that works out.

Reply:
I changed it once more, example bootstrap error:

```
[2023-04-21T15:07:20,388][ERROR][o.e.b.Elasticsearch      ] [hendrik-t14s] node validation exception
[1] bootstrap checks failed. You must address the points described in the following [1] lines before starting Elasticsearch.
bootstrap check failure [1] of [1]: Found an invalid configuration for xpack.ml.model_repository. If xpack.ml.model_repository is a file location, it must be placed below the configuration: file:///home/hendrik/work/test-install/elasticsearch-8.8.0-SNAPSHOT/config. See https://www.elastic.co/guide/en/machine-learning/8.8/ml-nlp-deploy-models.html#ml-nlp-deploy-model-air-gaped for more information.
```